src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,880 @@
+                };
+
+                let caller_mir = map.map.get_mut(&callsite.caller).unwrap();
+
+                let start = caller_mir.basic_blocks().len();
+
+                if !self.inline_call(callsite, caller_mir, callee_mir) {
+                    continue;
+                }
+
+                inlined_into.insert(callsite.caller);
+
+                // Add callsites from inlined function
+                for (bb, bb_data) in caller_mir.basic_blocks().iter_enumerated().skip(start) {
+                    // Only consider direct calls to functions
+                    let terminator = bb_data.terminator();
+                    if let TerminatorKind::Call {
+                        func: Operand::Constant(ref f), .. } = terminator.kind {
+                        if let ty::TyFnDef(callee_def_id, substs, _) = f.ty.sty {
+                            // Don't inline the same function multiple times.
+                            if callsite.callee != callee_def_id {
+                                callsites.push(CallSite {
+                                    caller: callsite.caller,
+                                    callee: callee_def_id,
+                                    substs: substs,
+                                    bb: bb,
+                                    location: terminator.source_info
+                                });
+                            }
+                        }
+                    }
+                }
+
+
+                csi -= 1;
+                if scc.len() == 1 {
+                    callsites.swap_remove(csi);
+                } else {
+                    callsites.remove(csi);
+                }
+
+                local_change = true;
+                changed = true;
+            }
+
+            if !local_change {
+                break;
+            }
+        }
+
+        // Simplify functions we inlined into.
+        for def_id in inlined_into {
+            let caller_mir = map.map.get_mut(&def_id).unwrap();
+            debug!("Running simplify cfg on {:?}", def_id);
+            CfgSimplifier::new(caller_mir).simplify();
+            remove_dead_blocks(caller_mir);
+        }
+        changed
+    }
+
+    fn get_foreign_mir(&mut self, def_id: DefId) -> Option<Rc<Mir<'tcx>>> {
+        if let Some(mir) = self.foreign_mirs.get(&def_id).cloned() {
+            return Some(mir);
+        }
+        // Cache the foreign MIR
+        let mir = self.tcx.sess.cstore.maybe_get_item_mir(self.tcx, def_id);
+        let mir = mir.map(Rc::new);
+        if let Some(ref mir) = mir {
+            self.foreign_mirs.insert(def_id, mir.clone());
+        }
+
+        mir
+    }
+
+    fn should_inline(&self, callsite: CallSite<'tcx>,
+                     callee_mir: &'a Mir<'tcx>) -> bool {
+
+        let tcx = self.tcx;
+
+        // Don't inline closures that have captures
+        // FIXME: Handle closures better
+        if callee_mir.upvar_decls.len() > 0 {
```

Comment:
I'm not sure how to handle the debuginfo related stuff, so I'm punting on it for now. You're right, but this isn't intended to be a complete implementation right away.